### PR TITLE
Implement quickcheck::Arbitrary for data types

### PR DESCRIPTION
**Pull request by emlun**
Alright, here's a go at #286. This adds a `quickcheck-gen` feature (it can't be just `quickcheck` because it also needs to enable `rand` - unless you require the user to also enable `rand` explicitly) and a subcrate `time-quickcheck-tests` that tests that feature.

I've intentionally left `Instant` out since it seems like values of that type aren't intrinsically meaningful on their own. It also seems quite easy to obtain an `Instant` from a `Duration` if needed, and that type does get an `Arbitrary` implementation.

### CODE REVIEW

**Comment by jhpratt** [comment]
I'll take a look at this when I get a chance.

Don't worry about the CI status right now; I had a change on the v0.3 branch that accidentally broke some things.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:57]: Two things:

- I don't think the type annotation is necessary here.
- It should be possible to just do `.try_into()`, given that the compiler knows an `i32` is expected (due to the `i32::max_value()`).

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:58]: Could we avoid single-letter variable names? It generally makes things easier to read.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:98]: A shim for the unstable `clamp` method would be _super_ useful for readability here. I can provide this if desired.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:100]: Directly constructing the `Duration` would be more efficient due to the checking of invariants in the constructor.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:74]: This could probably just be called `year`.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:239]: For `OffsetDateTime` — what would a user be expecting to happen during shrinking? I think (read: I haven't checked this) that this effectively shrinks the date, time, and offset, in that order. As the offset affects the calculated date/time, is this what a user would expect? I believe so, but I'm not sure :thinking:

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:261]: Having not tested this, is `shrink` called multiple times? From looking at the other types, it looks like a complete iterator is returned on the first call, where only a single item is returned here.

**Review by jhpratt — CHANGES_REQUESTED** [review]
For the tests, I just quickly glanced over things, as I don't particularly understand what's going on there.

If you rebase on the v0.3 branch, you should be able to trivially add `quickcheck-gen` to the `std,rand` group, which would add it to CI without doubling the number of checks run.

One thought about the license files: Should the author name be changed to "Jacob Pratt et al."? That would (correctly) indicate that I'm not the sole copyright owner.

Overall, things look good!

**Comment by emlun** [comment]
> If you rebase on the v0.3 branch, you should be able to trivially add `quickcheck-gen` to the `std,rand` group, which would add it to CI without doubling the number of checks run.

I'm not sure what you mean, I don't see anything like that in `Cargo.toml`. Maybe I'm not familiar enough with Cargo...

> One thought about the license files: Should the author name be changed to "Jacob Pratt et al."? That would (correctly) indicate that I'm not the sole copyright owner.

Sounds like a reasonable idea. Do you also want to do something to the `authors` fields in the `Cargo.toml` files? I don't think my contributions here are enough for me to be considered an author of the library proper, just wondering your thoughts on the `time-benchmarks/Cargo.toml` file for example.

> Overall, things look good!

Thanks!

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:261]: Yes, quickcheck uses `shrink` to iteratively minimize the test case, and it calls the method multiple times while doing so. I'm honestly not sure whether it's better to return "long" iterators at a time or just one, or how much it matters. I considered using `iter::successors` here but ultimately decided against it for no particular reason.

File: Cargo.toml
Comment (emlun) [reply Cargo.toml:37]: My main reason was to ensure that the `quickcheck-gen` feature is enabled, so it's impossible to accidentally not run the tests by not having the feature enabled. But I guess that doesn't really matter much since you run the tests against all the features in CI, so I can merge this into the base library if you prefer.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:57]: Yep, good catch!

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:55]: It is, but I personally prefer extracting the `where` clause to make the signature definition less busy. Do you prefer inlining the constraint?

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:58]: Sure. Should I do the same for the `Gen` parameters?

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:98]: Heheh, yeah, good idea! I can add one.

File: Cargo.toml
Comment (jhpratt) [reply Cargo.toml:37]: I've already got some other things gated in the tests, so gating this would be in line with that. Obviously it wouldn't be run unless the flag is enabled, but that's to be expected.

File: src/quickcheck.rs
Comment (jhpratt) [reply src/quickcheck.rs:55]: Yeah, I prefer it inlined. I believe that's actually how rustdoc shows it regardless.

File: src/quickcheck.rs
Comment (jhpratt) [reply src/quickcheck.rs:58]: Nah, since that's a "standard" parameter, if that makes sense. Sorta similar to using `T` for a type parameter.

File: src/quickcheck.rs
Comment (jhpratt) [reply src/quickcheck.rs:261]: If performance somehow becomes an issue with this, then it can be handled at that point. As long as the behavior is correct, it doesn't matter currently.

**Comment by jhpratt** [comment]
> > If you rebase on the v0.3 branch, you should be able to trivially add `quickcheck-gen` to the `std,rand` group, which would add it to CI without doubling the number of checks run.
> 
> I'm not sure what you mean, I don't see anything like that in `Cargo.toml`. Maybe I'm not familiar enough with Cargo...

This is in the `.github/workflows/build.yaml` file. There are a couple lines with `--group-features` there; I think it'll make more sense once you see that.

> > One thought about the license files: Should the author name be changed to "Jacob Pratt et al."? That would (correctly) indicate that I'm not the sole copyright owner.
> 
> Sounds like a reasonable idea. Do you also want to do something to the `authors` fields in the `Cargo.toml` files? I don't think my contributions here are enough for me to be considered an author of the library proper, just wondering your thoughts on the `time-benchmarks/Cargo.toml` file for example.

Adding people to the "authors" fields is presumably something that you just know when it's the time to do. Though adding a generic "time contributors" is never a bad idea.

With regard to the benchmarks, I believe you're already listed there!

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:100]: Yeah, I chose to use the constructor to prevent accidentally broken invariants. :slightly_smiling_face: But I'll change it.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:74]: Yep, agreed.

**Comment by emlun** [comment]
> > > If you rebase on the v0.3 branch, you should be able to trivially add `quickcheck-gen` to the `std,rand` group, which would add it to CI without doubling the number of checks run.
>
> This is in the `.github/workflows/build.yaml` file. There are a couple lines with `--group-features` there; I think it'll make more sense once you see that.

Oh, ok, did you mean the `serde,rand` group?

> With regard to the benchmarks, I believe you're already listed there!

Yeah, since I did write all of that sub-crate, I took the liberty of claiming authorship of that one. ^^ But I figured while we're on the subject I'd ask your thoughts on it. I don't know how likely it is that I'll remain the "main" author as it continues evolving in the future, but I guess it's just a matter of adding new ones when it seems appropriate.

File: src/quickcheck.rs
Comment (jhpratt) [reply src/quickcheck.rs:100]: Not an issue. But in cases like this where invariants are trivially proven, we may as well bypass the checks.

**Comment by jhpratt** [comment]
> Oh, ok, did you mean the serde,rand group?

Yeah, just looked over the previous comment and I'm not sure why I put that.

> Yeah, since I did write all of that sub-crate, I took the liberty of claiming authorship of that one. ^^ But I figured while we're on the subject I'd ask your thoughts on it. I don't know how likely it is that I'll remain the "main" author as it continues evolving in the future, but I guess it's just a matter of adding new ones when it seems appropriate.

Absolutely! If it ever truly matters (which realistically is unlikely), I'd just switch the order of the names. As it's not published on crates.io, it's less of an issue.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:239]: Yes, I believe it happens in that order - at least I had some issues in another project while implementing shrinking for a `BigRational` type. I had first shrunk the numerator and then the denominator (or maybe vice versa), which caused some issues when one of the two ended up many orders of magnitude greater than the other; reworking the shrinker to effectively alternate between them solved that issue.

My thinking here was to shrink the components in descending order of significance to traverse the value space efficiently, but I'm honestly not sure what a user would expect "shrinking a date" to do exactly. But I guess what matters most is the purpose of shrinking: to find minimal counterexamples. In that light I probably would as a user like to know if the particular offset is significant for the test failure (because it would be shrunk to zero if not). Ultimately I don't think the shrinking order matters much - you'll want the shrinker to generate likely continued failures first, but I don't think any of the components is more or less likely to be significant for any particular test failure.

Maybe it _would_ be more efficient to only generate one shrunk value per component per shrink step, though, so the shrinker doesn't have to run through N shrunk years and N shrunk days in each step before getting to the shrunk offsets, though. That would ultimately depend on the behaviour of `i32::shrink` etc., but I guess we could just limit that using `Iterator::take` anyway.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs]: Done. I left the closure parameters as is, though, as those are extremely local (usually within one line). Let me know if you'd like me to change those too.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs]: I ended up changing it before I saw your response. :innocent:

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs]: Done.

File: Cargo.toml
Comment (emlun) [reply Cargo.toml]: Done!

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs]: I see your thumbs up, does that mean you're happy with this as is or would you like me to limit the iterator sizes like I said in the last paragraph? :slightly_smiling_face:

Comment (jhpratt) [reply src/quickcheck.rs]: Yeah, it's fine as-is.

File: src/util.rs
Comment (jhpratt) [inline src/util.rs:101]: This should use `core` rather than `std`, and is why CI is failing.

File: src/util.rs
Comment (jhpratt) [inline src/util.rs:90]: As this isn't used anywhere aside from the quickcheck module, this trait can be placed there (with private visibility).

File: tests/quickcheck.rs
Comment (jhpratt) [inline tests/quickcheck.rs:1]: Is quickcheck `#![no_std]` compatible? If not, the reliance on `std` should be included `Cargo.toml` as well as here.

**Review by jhpratt — CHANGES_REQUESTED** [review]
Quick check looks good aside from a couple nits.

File: src/util.rs
Comment (emlun) [reply src/util.rs]: Done!

File: src/util.rs
Comment (emlun) [reply src/util.rs]: Done!

File: tests/quickcheck.rs
Comment (emlun) [reply tests/quickcheck.rs:1]: Since the `shrink` method returns a `std::boxed::Box`, I'm gonna say no. Fixed, thanks!

**Comment by jhpratt** [comment]
Also, just ran across [this post](https://blog.turbo.fish/cargo-features/) on Reddit. Perhaps it may be possible to use the `quickcheck` feature name after all, so long as the internal name is changed (which can ironically be changed right back with a simple `use` statement).

**Comment by emlun** [comment]
Nice find! That does indeed seem to work quite fine.

File: src/quickcheck.rs
Comment (jhpratt) [inline src/quickcheck.rs:242]:
```suggestion
        match g.gen_range(0, g.size().clamp(1, 7)) {
```

**Review by jhpratt — COMMENTED** [review]
These are minor enough that I can make these changes while cleaning up history. One is just a missed use of `clamp`, while the other is a complete hack that may or may not work.

File: src/quickcheck.rs
Comment (emlun) [reply src/quickcheck.rs:242]: Oh, good catch! Guess I missed that one because it wasn't spanning 5 lines or whatever... :laughing:

**Comment by jhpratt** [comment]
One question as I'm just about done with the history and verification: The implementation for `UtcOffset` appears to be able to generate `UtcOffset { seconds: -86_400 }` as values. Am I correct in this reading? If so, removing the `- 1` on the generation of the range would make this not the case, correct?

**Comment by jhpratt** [comment]
Ping @emlun — it's just the question in my previous comment that is blocking merging this.

**Comment by emlun** [comment]
Oh, sorry, I didn't notice. I'd say no, since the subtraction happens before the negation, so with `size = 86_400` the lower bound would be `-(86_400 - 1) = -86399`.

However, there _is_ indeed another bug here: with `size = 0`, the range ends up being `[1, 0)` which is invalid and causes a panic. I'll fix that.

**Comment by jhpratt** [comment]
I figured you just didn't see it — no big deal. I've just cleaned up history again, and will merge this pending CI.

**Review by jhpratt — APPROVED** [review]

**Comment by emlun** [comment]
Great, thanks! :smile:

**Pull request merged**
